File: examples/python/ch01_data_types/06_stream_consumer_group.py
"""Reference: docs/01-data-types/06-stream.md, docs/04-pubsub-streams/03-streams-consumer-group.md

Stream + Consumer Group. 두 컨슈머가 메시지를 분배 받아 처리.
"""
import sys, threading, time
from pathlib import Path

import redis  # noqa: F401

sys.path.insert(0, str(Path(__file__).parent.parent))
from _common import get_client, section  # noqa: E402

STREAM = "demo:tasks"
GROUP = "workers"


def consume(consumer_name: str, max_msgs: int = 5):
    r = get_client()
    consumed = 0
    while consumed < max_msgs:
        resp = r.xreadgroup(
            groupname=GROUP, consumername=consumer_name,
            streams={STREAM: ">"}, count=1, block=2000,
        )
        if not resp:
            break
        for _stream, entries in resp:
            for entry_id, fields in entries:
                print(f"[{consumer_name}] got {entry_id} :: {fields}")
                r.xack(STREAM, GROUP, entry_id)
                consumed += 1
    print(f"[{consumer_name}] consumed {consumed} done")


def main():
    r = get_client()
    r.delete(STREAM)

    try:
        r.xgroup_create(STREAM, GROUP, id="0", mkstream=True)
    except redis.exceptions.ResponseError:
        pass

    with section("프로듀서가 10개 메시지 추가"):
        for i in range(10):
            r.xadd(STREAM, {"task": f"job-{i}"})
        print("XLEN =", r.xlen(STREAM))

    with section("두 컨슈머 동시 실행 — 메시지 분배"):
        t1 = threading.Thread(target=consume, args=("worker-A", 5), daemon=True)
        t2 = threading.Thread(target=consume, args=("worker-B", 5), daemon=True)
        t1.start(); t2.start()
        t1.join(timeout=10); t2.join(timeout=10)

    with section("그룹 상태"):
        info = r.xinfo_groups(STREAM)
        for g in info:
            print(g)

    with section("정리"):
        r.delete(STREAM)
        print("done.")


if __name__ == "__main__":
    main()

File: examples/nodejs/ch01-data-types/06-stream-consumer-group.mjs
/**
 * Reference: docs/01-data-types/06-stream.md, docs/04-pubsub-streams/03-streams-consumer-group.md
 *
 * Stream + Consumer Group. 두 컨슈머가 메시지를 분배 받음.
 */
import { withClient, section, getClient } from "../_common.mjs";

const STREAM = "demo:tasks";
const GROUP = "workers";

async function consume(consumerName, maxMsgs) {
  const r = getClient();
  await r.connect();
  let consumed = 0;
  while (consumed < maxMsgs) {
    const resp = await r.xReadGroup(
      GROUP, consumerName,
      [{ key: STREAM, id: ">" }],
      { COUNT: 1, BLOCK: 2000 }
    );
    if (!resp || resp.length === 0) break;
    for (const stream of resp) {
      for (const msg of stream.messages) {
        console.log(`[${consumerName}] got ${msg.id} :: ${JSON.stringify(msg.message)}`);
        await r.xAck(STREAM, GROUP, msg.id);
        consumed++;
      }
    }
  }
  console.log(`[${consumerName}] consumed ${consumed} done`);
  await r.close();
}

await withClient(async (r) => {
  await r.del(STREAM);

  try {
    await r.xGroupCreate(STREAM, GROUP, "0", { MKSTREAM: true });
  } catch (_e) {
    // BUSYGROUP — 이미 존재
  }

  section("프로듀서가 10개 메시지 추가");
  for (let i = 0; i < 10; i++) {
    await r.xAdd(STREAM, "*", { task: `job-${i}` });
  }
  console.log("XLEN =", await r.xLen(STREAM));
});

section("두 컨슈머 동시 실행");
await Promise.all([consume("worker-A", 5), consume("worker-B", 5)]);

await withClient(async (r) => {
  section("그룹 상태");
  console.log(await r.xInfoGroups(STREAM));

  section("정리");
  await r.del(STREAM);
  console.log("done.");
});
